ospackage/recipes/artifacts.go

```diff
@@ -0,0 +1,186 @@
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package recipes
+
+import (
+	"context"
+	"crypto/sha256"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"path"
+	"path/filepath"
+	"runtime"
+	"strconv"
+	"strings"
+
+	"cloud.google.com/go/storage"
+	osconfigpb "github.com/GoogleCloudPlatform/osconfig/_internal/gapi-cloud-osconfig-go/google.golang.org/genproto/googleapis/cloud/osconfig/v1alpha2"
+)
+
+var (
+	testStorageClient *storage.Client
+	testHTTPClient    *http.Client
+)
+
+func newStorageClient(ctx context.Context) (*storage.Client, error) {
+	if testStorageClient != nil {
+		return testStorageClient, nil
+	}
+	return storage.NewClient(ctx)
+}
+
+func newHTTPClient() *http.Client {
+	if testHTTPClient != nil {
+		return testHTTPClient
+	}
+	return &http.Client{}
+}
+
+// FetchArtifacts takes in a slice of artifacs and dowloads them into the specified directory,
+// Returns a map of artifact names to their new locations on the local disk.
+func FetchArtifacts(ctx context.Context, artifacts []*osconfigpb.SoftwareRecipe_Artifact, directory string) (map[string]string, error) {
+	localNames := make(map[string]string)
+
+	for _, a := range artifacts {
+		path, err := fetchArtifact(ctx, a, directory)
+		if err != nil {
+			return nil, err
+		}
+		localNames[a.Id] = path
+	}
+
+	return localNames, nil
+}
+
+func fetchArtifact(ctx context.Context, artifact *osconfigpb.SoftwareRecipe_Artifact, directory string) (string, error) {
+	localPath := path.Join(directory, artifact.Id)
+	uri, err := url.Parse(artifact.Uri)
+	if err != nil {
+		return "", fmt.Errorf("Could not parse url %q for artifact %q", artifact.Uri, artifact.Id)
+	}
+
+	var reader io.ReadCloser
+	switch strings.ToLower(uri.Scheme) {
+	case "gcs":
+		reader, err = fetchWithGCS(ctx, uri.Host, uri.Path, uri.Fragment)
+		defer reader.Close()
+		if err != nil {
+			return "", fmt.Errorf("error fetching artifact %q from GCS: %v", artifact.Id, err)
+		}
+	case "https", "http":
+		reader, err = fetchWithHTTP(ctx, artifact.Uri)
+		defer reader.Close()
+		if err != nil {
+			return "", fmt.Errorf("error fetching artifact %q with http or https: %v", artifact.Id, err)
+		}
+	default:
+		return "", fmt.Errorf("artifact %q has unsupported protocol %s", artifact.Id, uri.Scheme)
+	}
+
+	err = downloadStream(reader, artifact.Checksum, localPath)
+	if err != nil {
+		return "", err
+	}
+	return localPath, nil
+}
+
+func fetchWithGCS(ctx context.Context, bucket, path, generation string) (io.ReadCloser, error) {
+	client, err := newStorageClient(ctx)
```

Comment:
We should think about adding a fallback for unauthenticated access in case we don't have a service account to use.

Reply:
how do I detect that / what would a fallback look like?

Reply:
Isn't that what happens here? The storage client is created with the current context which may not have any auth. I'd imagine that calling the API will fail the same way if there's no SA and the object requires access or if there is an SA but the account doesn't have access to the object.

Reply:
The fallback would be to create the URL and download it via the HTTP function

Reply:
@adjackura As per Ryan's question is that very different than what the storage client will do if there isn't a service account. I'm having trouble finding out what it does if there isn't a service account but my assumption is that it would fallback to no authentication.

Also I found that the storage client has a WithoutAuthentication option so if we need to be explicit about not having authentication it looks like we could still use the storage client for retrieval.